src/vs/platform/terminal/node/ptyService.ts

```diff
@@ -84,27 +121,302 @@ export class PtyService extends Disposable implements IPtyService {
+
+	private async _expandTerminalTab(tab: ITerminalTabLayoutInfoById): Promise<ITerminalTabLayoutInfoDto> {
+		const expandedTerminals = (await Promise.all(tab.terminals.map(t => this._expandTerminalInstance(t))));
+		const filtered = expandedTerminals.filter(term => term.terminal !== null) as IRawTerminalInstanceLayoutInfo<IPtyHostDescriptionDto>[];
+		return {
+			isActive: tab.isActive,
+			activeTerminalProcessId: tab.activeTerminalProcessId,
+			terminals: filtered
+		};
+	}
+
+	private async _expandTerminalInstance(t: ITerminalInstanceLayoutInfoById): Promise<IRawTerminalInstanceLayoutInfo<IPtyHostDescriptionDto | null>> {
+		const persistentTerminalProcess = this._throwIfNoPty(t.terminal);
+		const termDto = persistentTerminalProcess && await this._terminalToDto(t.terminal, persistentTerminalProcess);
+		return {
+			terminal: termDto ?? null,
+			relativeSize: t.relativeSize
+		};
+	}
+
+	private async _terminalToDto(id: number, persistentTerminalProcess: PersistentTerminalProcess): Promise<IPtyHostDescriptionDto> {
+		const [cwd, isOrphan] = await Promise.all([persistentTerminalProcess.getCwd(), persistentTerminalProcess.isOrphaned()]);
+		return {
+			id,
+			title: persistentTerminalProcess.title,
+			pid: persistentTerminalProcess.pid,
+			workspaceId: persistentTerminalProcess.workspaceId,
+			workspaceName: persistentTerminalProcess.workspaceName,
+			cwd,
+			isOrphan
+		};
+	}
+
+	private _throwIfNoPty(id: number): PersistentTerminalProcess {
 		const pty = this._ptys.get(id);
 		if (!pty) {
 			throw new Error(`Could not find pty with id "${id}"`);
 		}
 		return pty;
 	}
+
+	// private _orphanQuestionReply(args: IOrphanQuestionReplyArgs): void {
+	// 	const persistentTerminalProcess = this._throwIfNoPty(args.id);
+	// 	persistentTerminalProcess.orphanQuestionReply();
+	// }
+}
+
+export class PersistentTerminalProcess extends Disposable {
+
+	// private readonly _bufferer: TerminalDataBufferer;
+
+	private readonly _pendingCommands = new Map<number, { resolve: (data: any) => void; reject: (err: any) => void; }>();
+
+	private readonly _recorder: TerminalRecorder;
+	private _isStarted: boolean = false;
+
+	private _orphanQuestionBarrier: AutoOpenBarrier | null;
+	private _orphanQuestionReplyTime: number;
+	private _orphanRequestQueue = new Queue<boolean>();
+	private _disconnectRunner1: RunOnceScheduler;
+	private _disconnectRunner2: RunOnceScheduler;
+	private readonly _onProcessReplay = this._register(new Emitter<IPtyHostProcessReplayEvent>());
+	public get onProcessReplay(): Event<IPtyHostProcessReplayEvent> { return this._onProcessReplay.event; }
+	private readonly _onProcessReady = this._register(new Emitter<{ pid: number, cwd: string }>());
+	public get onProcessReady(): Event<{ pid: number, cwd: string }> { return this._onProcessReady.event; }
+	private readonly _onProcessTitleChanged = this._register(new Emitter<string>());
+	public get onProcessTitleChanged(): Event<string> { return this._onProcessTitleChanged.event; }
```

Comment:
Let's switch this to readonly like the others, also prefer to leave public off for these